README.md
Feedybacky
==========

### What is Feedybacky? ###

Feedybacky is simple JS plugin for web pages which facilitates feedback system. It creates a sliding element on HTML page in which a user can provide a request (error, suggest, comment etc.) and send it in rapid and easy way. The special part of this process is the fact, that Feedybacky allows to attach a screenshot of the current page and additional information **automatically**. It includes i.a. current URL, user agent string and - of course - current datetime. In many situations, request's text provided by a user is not sufficient to reproduce e.g. an error - Feedybacky tries to solve that problem.

Feedybacky uses `html2canvas` library created by Niklas von Hertzen.

### Installation ###

If you just want to use Feedybacky, you can copy the ZIP archive from the repository and unzip to the destinated directory. It is important to have all subfolders (`css`, `js` etc.) next to each other in the same folder.

Of course, you can also install Feedybacky as NPM package:

	npm install --save feedybacky

If you want to modify or build Feedybacky by yourself, you should install NPM dependencies by running `npm install` command, install `gulp-cli` globally (`npm install -g gulp-cli`) and run `gulp` in Feedybacky directory (`npm install gulp`). Afterwards, you can copy following directories to the destinated folder: `css`, `dependencies`, `i18n`, `img`, `js`.

Lastly, you can also fetch Feedybacky via CDN - if you would like to use the plugin this way, please check example page `examples/09-cdn.html` to display an example.

### How to include Feedybacky in my web page? ###

#### Standard way ####

Firstly, you have to create a DOM element anywhere on your web page:
```html
<div id="feedybacky-container"></div>
```	
Secondly, Feedybacky script should be included and its object initialized:
```javascript
<script type="text/javascript" src="js/feedybacky.min.js"></script>
//...
var feedybacky = new Feedybacky('feedybacky-container', {
	onSubmitUrl: url
});
```

If you want to pair plugin with Feedybacky portal, you need to fetch your Feedybacky API key and create a project in the portal to retrieve its symbol. Then you can use it on your WWW site in order to send request to Feedybacky portal. You should also provide URLs to terms and conditions and privacy policy of site, on which you place the plugin. The "legal" checkboxes are not checked by default but if a user checks them, the information in local storage is saved so these checkboxes would be checked by default later.
```javascript
var feedybacky = new Feedybacky('feedybacky-container', {
	apiKey: 'your Feedybacky API key',
	projectSymbol: 'your Feedybacky project symbol',
	termsUrl: 'link to terms and conditions',
	privacyPolicyUrl: 'link t privacy policy'
});
```
#### Angular ####

1. Add a DOM element to your root `.html` template (outside `<app-root></app-root>`):
```html
<div id="feedybacky-container"></div>
```
2. Edit your `angular.json` file:
```json
//...
"build": {
	//...
	{
		"styles": [
			//...
			"node_modules/feedybacky/css/feedybacky.min.css"
		],
		"scripts": [
			//...
			"node_modules/feedybacky/dependencies/html2canvas/html2canvas.min.js"
		]
	}
}
```
3. In your `app.component.ts` add:
```typescript
import { Feedybacky } from 'feedybacky';

export class AppComponent {
	protected feedybacky = new Feedybacky('feedybacky-container', {
		onSubmitUrl: url
	});
	//...
}
```
#### Vue.js ####
1. Add a DOM element to `public/index.html`:
```html
<div id="feedybacky-container"></div>
```
2. Import Feedybacky into `App.vue`:
```ts
import { Feedybacky } from 'feedybacky'

const feedybacky = new Feedybacky('feedybacky-container', {
  onSubmitUrl: url
});
```
3. Import additional styles and dependencies in `main.js`:
```ts
import 'feedybacky/css/feedybacky.min.css';
import 'feedybacky/dependencies/html2canvas/html2canvas.min.js';
```

#### React ####
1. Add DOM element to `public/index.html`:
```html
<div id="feedybacky-container"></div>
```
2. Import and init Feedybacky and import dependencies into `App.js`
```ts
import { Feedybacky } from 'feedybacky';
import 'feedybacky/css/feedybacky.min.css';
import 'feedybacky/dependencies/html2canvas/html2canvas.min.js';

new Feedybacky('feedybacky-container', {
  onSubmitUrl: url
})
```

### Parameters ###

In Feedybacky constructor, next to ID of empty div, a JSON object with parameters should be provided. Below there are parameters accepted by Feedybacky.

`language` - language in which Feedybacky should display default messages. Can be omitted, especially if custom messages are provided. Available language packs with default messages can be found in `i18n` directory. Default value: `"en"`

`onSubmit` - callback function to invoke after sending a request. Either `onSubmit` or `onSubmitUrl` parameter must be set. This parameter is for advanced usage only - in most cases `onSubmitUrl` would be the better choice.

`onSubmitUrl` - URL with POST web service to receive information about a sent request. The resulted JSON object can consist of following fields:

* **message** - text or recording provided by the user.
* **messageType** - type of message (`text` or `voice`).
* **timestamp** - date and time of the request creation.
* **url** - URL on which the request is sent.
* **errors** - array with error messages from JavaScripts's console.
* **image** - screenshot of the current web page. Available only if the user accepted sending a screenshot.
* **agent** - user agent string. Available only if the user accepted sending metadata information.
* **cookies** - current user's cookie string. Available only if the user accepted sending metadata information.
* **platform** - platform symbol of the user's machine. Available only if the user accepted sending metadata information.
* **screenSize** - size of the screen (result of `screen.width` and `screen.height`). Available only if the user accepted sending metadata information.
* **availableScreenSize** - size of available part of the screen (result of `screen.availWidth` and `screen.availHeight`). Available only if the user accepted sending metadata information.
* **innerSize** - size of visible frame (result of `window.innerWidth` and `window.innerHeight`). Available only if the user accepted sending metadata information.
* **colorDepth** - depth of the pixel on the screen (result of `screen.colorDepth`). Available only if the user accepted sending metadata information.
* **orientation** - screen orientation (result of `screen.orientation.type;`). Available only if the user accepted sending metadata information.
* **cookieEnabled** - information about if a user enabled cookies (result of `navigator.cookieEnabled`). Available only if the user accepted sending metadata information.
* **browserLanguage** - language in which a browser displays the site (result of `navigator.language`). Available only if the user accepted sending metadata information.
* **referrer** - URL referrer from which the user comes to the site (result of `document.referrer`). Available only if the user accepted sending metadata information.
* **pixelRatio** - ratio of device physical pixels to CSS pixels (result of `window.devicePixelRatio`). Available only if the user accepted sending metadata information.
* **offsetX** - horizontally scrolled pixels (result of `window.pageXOffset`). Available only if the user accepted sending metadata information.
* **offsetY** - vertically scrolled pixels (result of `window.pageYOffset`). Available only if the user accepted sending metadata information.
* **history** - array of objects representing last N events called by the user (details below). Available only if the user accepted sending history information.
* **extraInfo** - JSON object with extra parameters passed by the callback function. Available only if extra data callback was defined.
* **email** - e-mail address provided by the user if available (the appropriate field is visible).
* **category** - value of the category selected by the user if available (the appropriate field is visible).
* **priority** - value of priority ("medium" or "high") selected by the user if available (the appropriate field is visible).
* **prefix** - prefix defined for the Feedybacky instance (if defined).
* **adBlock** - information if an enabled AdBlock browser plugin can be detected. It can have value 1 or 0.
* **visitedUrls** - array of objects with last N visited URLs withing the site. Available only if the parameter `urlTracking` is enabled.

`apiKey` - API key generated for your Feedybacky account in the portal. You can retrieve it from the portal on the edit account view under the "API key" section. Setting this parameter is obligatory if you want to send request to the Feedybacky portal. If this and `projectSymbol` are set, the value of `onSubmitUrl` is replaced by the official Feedybacky portal endpoint.

`projectSymbol` - the symbol of a project created in Feedybacky portal. You can fetch it from the portal on the project config view. Setting this parameter is obligatory if you want to send request to the Feedybacky portal. If this and `apiKey` are set, the value of `onSubmitUrl` is replaced by the official Feedybacky portal endpoint.

`termsUrl` - URL for terms and conditions document of a site. It is used when the plugin would send the request to official Feedybacky portal. Link would be opened in a new tab. Default value is `#`.

`privacyPolicyUrl` - URL for privacy policy document of a site. It is used when the plugin would send the request to official Feedybacky portal. Link would be opened in a new tab. Default value is `#`.

`texts` - JSON object with custom messages in different parts of the plugin. It can contain following keys:

* **tooltip** - information visible after placing the cursor on the minified div with the plugin.
* **title** - title on top of the plugin screen.
* **description** - description on the plugin screen (before textarea).
* **additionalDataInformation** - description on the plugin screen (after textarea, before checkboxes).
* **screenshot** - label for the checkbox to accept sending a screenshot.
* **metadata** - label for the checkbox to accept sending metadata.
* **history** - label for the checkbox to accept sending event history.
* **termsAccepted** - label for the checkbox to accept terms and conditions (visible when the plugin is paired with the Feedybacky portal).
* **personalDataAccepted** - label for the checkbox to confirm personal data processing (visible when the plugin is paired with the Feedybacky portal).
* **note** - label for the text between checkboxes and send button (empty by default and then not visible).
* **send** - label for the "Send" button.
* **requestSuccess** - message visible after successful POST request sending.
* **requestFail** - message visible after unsuccessful POST request sending.
* **requestPending** - message visible during sending a POST request.
* **powered** - message on bottom of the plugin screen.
* **error403** - message after a situation where an 403 error occurred.
* **error404** - message after a situation where an 404 error occurred.
* **error500** - message after a situation where an 500 error occurred.
* **descriptionErrorEmpty** - error message for the empty description field.
* **email** - label before the e-mail address field.
* **emailErrorEmpty** - error message for the empty e-mail address field.
* **termsAcceptedErrorNotChecked** - error message when terms and conditions have been not accepted while the plugin is paired with the Feedybacky portal.
* **personalDataAcceptedErrorNotChecked** - error message when user has not confirm information about personal data processing while the plugin is paired with the Feedybacky portal.
* **category** - label before the category field.
* **priority** - label before the priority field.
* **modifyScreenshot** - label for the link for screenshot modification.
* **working** - label displayed when work is in progress (during modified screenshot generation).
* **messageTypeText** - label for text message type.
* **messageTypeVoice** - label for voice message type.
* **voiceRecord** - label for record start button (for voice message type).
* **voiceStopRecord** - label for record stop button (for voice message type).
* **voicePlay** - label for record play button (for voice message type).
* **voiceRecordTitle** - label for record start button title (for voice message type).
* **voiceStopRecordTitle** - label for record stop button title (for voice message type).
* **voicePlayTitle** - label for record play button title (for voice message type).
* **descriptionErrorVoiceTooLong** - error message when recording for voice message type is too long.

`extraInfo` - optional callback function with no parameter which only returns JSON object. Keys and values of the object are merged with standard request information (next to "message", "timestamp" etc.). It can be used to pass extra data specific to the web application, such as user ID.

`emailField` - optional parameter indicating if e-mail address field should be visible. If it is visible, it is also required. Default value: `false`.

`categories` - optional parameter for array of category options presented for the user. If this parameter is omitted or no category is provided, the select with category is not visible and the value not be sent with a request. Each element of this array should be a JSON object with two keys: `label` and `value`, e.g.:

```ts
categories: [
	{ label: 'Suggestion', value: 'suggest' },
	{ label: 'Issue', value: 'error' },
],
```

`priorityField` - optional parameter indicating if priority select should be visible. Default value: `false`.

`screenshotField ` - optional parameter for defining behaviour of the plugin during processing a screenshot. The default value is `"visible"` and it means that the checkbox is visible and a user can select if they would like to send a screenshot or not. The other possible values are `"autoEnable"` (the checkbox is not visible and the screenshot is sent automatically) and `"autoDisable"` (the checkbox is not visible and the screenshot is ignored).

`metadataField` - optional parameter for defining behaviour of the plugin during processing metadata. The default value is `"visible"` and other possibilities are similar as for `screenshotField` above.


changelog.txt
Version 2.3.1 (2023-09-17)
========================

Security fix - hiding values of inputs with type "password" in the history field (there are replaced for "***").

Version 2.3 (2022-02-12)
========================

1. Modifying a screenshot and create markups on it is possible.
2. Message types introduced - voice is a new possibility to provide a request's message.
3. Refactoring Gulpfile and method to minimize and uglify JS and CSS files.

Version 2.2 (2021-08-23)
========================

1. Setting themes and predefined dark theme.
2. Additional metadata information: cookieEnabled, browserLanguage, referrer, pixelRatio, offsetX, offsetY.
3. Selecting a category.
4. Selecting a priority (among high and medium).
5. Selecting a screenshot method (among default html2canvas and mediaSource).
6. Minor fixes: error 403 message, changes in labels.

Version 2.1 (2021-07-01)
========================

1. Introducing URL tracking.
2. Fixed and adjusted Wordpress wrapper.
3. Slightly improved responsiveness - the plugin is scrollable, wider and has smaller font size on more narrowed screens.

Version 2.0.1 (2021-06-03)
==========================

Small bug fixes.

Version 2.0 (2021-06-03)
========================

1. Params for official integration introduced - "apiKey", "projectSymbol", "termsUrl" and "privacyPolicyUrl".
2. When the plugin is paired with official Feedybacky portal, the additional legal checkboxes are displayed.
3. During sending a request, the message about operation is displayed.
4. Fixed bug - "emailField" and "expandMessageLink " now are interpreted correctly if "false" would be typed as string.
5. Fixed bug - after sending a request, a site is not scrolled to the top automatically.
6. Official logo of Feedybacky has been changed.
7. Default CSS adjusted in order to display more checkboxes properly.

Version 1.5 (2020-03-14)
========================

1. New section can be added to the form - note, located between checkboxes and send button by default.
2. Reordering elements inside the form is possible.
3. Feedybacky can be located on the left side.
4. Most of Feedybacky's elements can have additional class in order to further style (classes object).
5. Expand link for message can be added.

Version 1.4 (2020-02-22)
========================

1. It is possible to save history of some user actions (basing on set of JavaScript's event types) and send last N actions with request.
2. The form can be opened or closed programatically with methods "open" and "close".
3. Basic detection of enabled AdBlock has been added with the possibility of passing the value test from a external source.

Version 1.3 (2020-02-16)
========================

1. Callback "beforeSubmit" can be defined.
2. Callback "onSubmitUrlSuccess" can be defined.
3. Callback "onSubmitUrlError" can be defined.
4. Prefix can be defined for the form.
5. Alerts after request can be enabled or disabled by "alertAfterRequest" parameter.
6. Fix with checkbox display on the form.
7. Directory with examples is present.

Version 1.2 (2019-12-27)
========================

1. E-mail address field can be added to the form.
2. Checkboxes for screenshot and metadata can be hidden and automatically passed or ignored.

Version 1.1 (2019-12-07)
========================

1. Description of the request is required.
2. Extra info object can be passed with the request.





wrappers/wordpress/feedybacky/feedybacky.php
<?php
/**
 * Plugin Name: Feedybacky
 * Plugin URI: https://github.com/WildaSoftware/Feedybacky
 * Description: Feedybacky's plugin wrapper for Wordpress. The plugin can be used to facilitate feedback gathering from a web site. This version based on version 2.1 of the original plugin.
 * Version: 1.1
 * Author: Wilda Software
 * Author URI: https://wildasoftware.pl/
 **/

const FYBY_OPTION__LANGUAGE = 'fyby_option__language';
const FYBY_OPTION__ONSUBMIT = 'fyby_option__onsubmit';
const FYBY_OPTION__ONSUBMITURL = 'fyby_option__onsubmiturl';
const FYBY_OPTION__EXTRAINFO = 'fyby_option__extrainfo';
const FYBY_OPTION__EMAILFIELD = 'fyby_option__emailfield';
const FYBY_OPTION__SCREENSHOTFIELD = 'fyby_option__screenshotfield';
const FYBY_OPTION__METADATAFIELD = 'fyby_option__metadatafield';
const FYBY_OPTION__HISTORYFIELD = 'fyby_option__historyfield';
const FYBY_OPTION__HISTORYLIMIT = 'fyby_option__historylimit';
const FYBY_OPTION__BEFORESUBMIT = 'fyby_option__beforesubmit';
const FYBY_OPTION__ONSUBMITURLSUCCESS = 'fyby_option__onsubmiturlsuccess';
const FYBY_OPTION__ONSUBMITURLERROR = 'fyby_option__onsubmiturlerror';
const FYBY_OPTION__PREFIX = 'fyby_option__prefix';
const FYBY_OPTION__ALERTAFTERREQUEST = 'fyby_option__alertafterrequest';
const FYBY_OPTION__ADBLOCKDETECTED = 'fyby_option__adblockdetected';
const FYBY_OPTION__SIDE = 'fyby_option__side';
const FYBY_OPTION__ORDER = 'fyby_option__order';
const FYBY_OPTION__EXPANDMESSAGELINK = 'fyby_option__expandmessagelink';
const FYBY_OPTION__TEXTS = 'fyby_option__texts';
const FYBY_OPTION__CLASSES = 'fyby_option__classes';
const FYBY_OPTION__APIKEY = 'fyby_option__apikey';
const FYBY_OPTION__PROJECTSYMBOL = 'fyby_option__projectsymbol';
const FYBY_OPTION__TERMSURL = 'fyby_option__termsurl';
const FYBY_OPTION__PRIVACYPOLICYURL = 'fyby_option__privacypolicyurl';
const FYBY_OPTION__URLTRACKING = 'fyby_option__urltracking';
const FYBY_OPTION__URLTRACKINGLIMIT = 'fyby_option__urltrackinglimit';
const FYBY_OPTION__THEME = 'fyby_option__theme';
const FYBY_OPTION__ALLOWEDTHEMES = 'fyby_option__allowedthemes';
const FYBY_OPTION__CATEGORIES = 'fyby_option__categories';
const FYBY_OPTION__PRIORIRYFIELD = 'fyby_option__priorityfield';
const FYBY_OPTION__SCREENSHOTMETHOD = 'fyby_option__screenshotmethod';

function fyby__setup_menu() {
	add_options_page('Feedybacky', 'Feedybacky', 'manage_options', 'feedybacky-config', 'fyby__handle_menu_view' );
}

function fyby__generate_input($name, $param, $type, $label, $options = []) {
	$field = '<tr><th scope="row"><label for="FeedybackyConfig['.$name.']">'.$label.'</label></th><td>';
	
	if($type == 'text') {
		$field .= '
			<input type="text" id="feedybacky-config-'.strtolower($name).'" name="FeedybackyConfig['.$name.']" value="'.get_option($param).'" class="regular-text">
		';
	}
	elseif($type == 'textarea') {
		$field .= '
			<textarea id="feedybacky-config-'.strtolower($name).'" form="feedybacky-config" name="FeedybackyConfig['.$name.']" class="regular-text">'.get_option($param).'</textarea>
		';
	}
	elseif($type == 'checkbox') {
		$field .= '
			<input type="checkbox" id="feedybacky-config-'.strtolower($name).'" name="FeedybackyConfig['.$name.']" '.(!empty(get_option($param)) ? 'checked' : '').'>
		';
	}
	elseif($type == 'select') {
		$paramValue = get_option($param);
		
		$text = '
			<select id="feedybacky-config-'.strtolower($name).'" name="FeedybackyConfig['.$name.']">
		';
		
		foreach($options as $optionValue => $optionLabel) {
			$text .= '<option value="'.$optionValue.'" '.($paramValue == $optionValue ? 'selected' : '').'>'.$optionLabel.'</option>';
		}
		
		$text .= '</select>';
		
		$field .= $text;
	}
	elseif($type == 'number') {
		$field .= '
			<input type="number" id="feedybacky-config-'.strtolower($name).'" name="FeedybackyConfig['.$name.']" value="'.get_option($param).'" min="0">
		';
	}
	
	$field .= '</td></tr>';
	
	echo $field;
}
 
function fyby__handle_menu_view(){
	if(!empty($_POST['FeedybackyConfig'])) {
		$data = $_POST['FeedybackyConfig'];		
		update_option(FYBY_OPTION__LANGUAGE, $data['language']);
		update_option(FYBY_OPTION__ONSUBMIT, str_replace('\\', '', $data['onSubmit']));
		update_option(FYBY_OPTION__ONSUBMITURL, $data['onSubmitUrl']);
		update_option(FYBY_OPTION__APIKEY, $data['apiKey']);
		update_option(FYBY_OPTION__PROJECTSYMBOL, $data['projectSymbol']);
		update_option(FYBY_OPTION__TERMSURL, $data['termsUrl']);
		update_option(FYBY_OPTION__PRIVACYPOLICYURL, $data['privacyPolicyUrl']);
		update_option(FYBY_OPTION__EXTRAINFO, str_replace('\\', '', $data['extraInfo']));
		update_option(FYBY_OPTION__EMAILFIELD, !empty($data['emailField']));
		update_option(FYBY_OPTION__SCREENSHOTFIELD, $data['screenshotField']);
		update_option(FYBY_OPTION__METADATAFIELD, $data['metadataField']);
		update_option(FYBY_OPTION__HISTORYFIELD, $data['historyField']);
		update_option(FYBY_OPTION__HISTORYLIMIT, $data['historyLimit']);
		update_option(FYBY_OPTION__BEFORESUBMIT, str_replace('\\', '', $data['beforeSubmit']));
		update_option(FYBY_OPTION__ONSUBMITURLSUCCESS, str_replace('\\', '', $data['onSubmitUrlSuccess']));
		update_option(FYBY_OPTION__ONSUBMITURLERROR, str_replace('\\', '', $data['onSubmitUrlError']));
		update_option(FYBY_OPTION__PREFIX, $data['prefix']);
		update_option(FYBY_OPTION__ALERTAFTERREQUEST, !empty($data['alertAfterRequest']));
		update_option(FYBY_OPTION__ADBLOCKDETECTED, $data['adBlockDetected']);
		update_option(FYBY_OPTION__SIDE, $data['side']);
		update_option(FYBY_OPTION__ORDER, $data['order']);
		update_option(FYBY_OPTION__EXPANDMESSAGELINK, !empty($data['expandMessageLink']));
		update_option(FYBY_OPTION__TEXTS, str_replace('\\', '', $data['texts']));
		update_option(FYBY_OPTION__CLASSES, str_replace('\\', '', $data['classes']));
		update_option(FYBY_OPTION__URLTRACKING, !empty($data['urlTracking']));
		update_option(FYBY_OPTION__URLTRACKINGLIMIT, $data['urlTrackingLimit']);
		update_option(FYBY_OPTION__THEME, $data['theme']);
		update_option(FYBY_OPTION__ALLOWEDTHEMES, str_replace('\\', '', $data['allowedThemes']));
		update_option(FYBY_OPTION__CATEGORIES, str_replace('\\', '', $data['categories']));
		update_option(FYBY_OPTION__PRIORIRYFIELD, $data['priorityField']);
		update_option(FYBY_OPTION__SCREENSHOTMETHOD, $data['screenshotMethod']);
	}
	
	echo '<div id="feedybacky-config-page">';
	 
	echo '<img id="feedybacky-admin-logo" src="'.WP_PLUGIN_URL.'/feedybacky/assets/logo_feedybacky.png"/><div id="feedybacky-admin-header" style="font-weight: bold; font-size: 1.2rem">Plugin config</div>';
	echo '<form id="feedybacky-config" method="post">';
	
	echo '<table class="form-table" role="presentation">';
	echo '<tbody>';
	
	echo fyby__generate_input('language', FYBY_OPTION__LANGUAGE, 'text', 'Language');
	echo fyby__generate_input('onSubmit', FYBY_OPTION__ONSUBMIT, 'textarea', 'On Submit');
	echo fyby__generate_input('onSubmitUrl', FYBY_OPTION__ONSUBMITURL, 'text', 'On Submit URL');
	echo fyby__generate_input('apiKey', FYBY_OPTION__APIKEY, 'text', 'API key');
	echo fyby__generate_input('projectSymbol', FYBY_OPTION__PROJECTSYMBOL, 'text', 'Project symbol');
	echo fyby__generate_input('termsUrl', FYBY_OPTION__TERMSURL, 'text', 'URL to terms and conditions');
	echo fyby__generate_input('privacyPolicyUrl', FYBY_OPTION__PRIVACYPOLICYURL, 'text', 'URL to privacy policy');
	echo fyby__generate_input('extraInfo', FYBY_OPTION__EXTRAINFO, 'textarea', 'Extra Info');
	echo fyby__generate_input('emailField', FYBY_OPTION__EMAILFIELD, 'checkbox', 'Email Field');
	echo fyby__generate_input('screenshotField', FYBY_OPTION__SCREENSHOTFIELD, 'select', 'Screenshot Field', ['visible' => 'Visible', 'autoEnable' => 'Auto Enable','autoDisable' => 'Auto Disable']);
	echo fyby__generate_input('metadataField', FYBY_OPTION__METADATAFIELD, 'select', 'Metadata Field', ['visible' => 'Visible', 'autoEnable' => 'Auto Enable', 'autoDisable' => 'Auto Disable']);
	echo fyby__generate_input('historyField', FYBY_OPTION__HISTORYFIELD, 'select', 'History Field', ['visible' => 'Visible', 'autoEnable' => 'Auto Enable', 'autoDisable' => 'Auto Disable']);
	echo fyby__generate_input('historyLimit', FYBY_OPTION__HISTORYLIMIT, 'number', 'History Limit');
	echo fyby__generate_input('urlTracking', FYBY_OPTION__URLTRACKING, 'checkbox', 'URL Tracking');
	echo fyby__generate_input('urlTrackingLimit', FYBY_OPTION__URLTRACKINGLIMIT, 'number', 'URL Tracking Limit');
	echo fyby__generate_input('beforeSubmit', FYBY_OPTION__BEFORESUBMIT, 'textarea', 'Before Submit');
	echo fyby__generate_input('onSubmitUrlSuccess', FYBY_OPTION__ONSUBMITURLSUCCESS, 'textarea', 'On Submit Url Success');
	echo fyby__generate_input('onSubmitUrlError', FYBY_OPTION__ONSUBMITURLERROR, 'textarea', 'On Submit Url Error');
	echo fyby__generate_input('prefix', FYBY_OPTION__PREFIX, 'text', 'Prefix');
	echo fyby__generate_input('alertAfterRequest', FYBY_OPTION__ALERTAFTERREQUEST, 'checkbox', 'Alert After Request');
	echo fyby__generate_input('adBlockDetected', FYBY_OPTION__ADBLOCKDETECTED, 'text', 'AdBlock Detected');
	echo fyby__generate_input('side', FYBY_OPTION__SIDE, 'text', 'Side');
	echo fyby__generate_input('order', FYBY_OPTION__ORDER, 'text', 'Order');
	echo fyby__generate_input('expandMessageLink', FYBY_OPTION__EXPANDMESSAGELINK, 'checkbox', 'Expand Message Link');
	echo fyby__generate_input('texts', FYBY_OPTION__TEXTS, 'textarea', 'Texts');
	echo fyby__generate_input('classes', FYBY_OPTION__CLASSES, 'textarea', 'Classes');
	echo fyby__generate_input('theme', FYBY_OPTION__THEME, 'text', 'Theme');
	echo fyby__generate_input('allowedThemes', FYBY_OPTION__ALLOWEDTHEMES, 'textarea', 'Allowed Themes');
	echo fyby__generate_input('categories', FYBY_OPTION__CATEGORIES, 'textarea', 'Categories');
	echo fyby__generate_input('priorityField', FYBY_OPTION__PRIORIRYFIELD, 'checkbox', 'Priority Field');
	echo fyby__generate_input('screenshotMethod', FYBY_OPTION__SCREENSHOTMETHOD, 'text', 'Screenshot Method');
	
	echo '</tbody>';
	echo '</table>';
	
	echo submit_button("Save");
	echo '</form>';
	
	echo '</div>';
}

function fyby__init_plugin() {
	if(empty(get_option(FYBY_OPTION__LANGUAGE))) {
		update_option(FYBY_OPTION__LANGUAGE, 'en');
	}

	if(empty(get_option(FYBY_OPTION__SCREENSHOTFIELD))) {
		update_option(FYBY_OPTION__SCREENSHOTFIELD, 'visible');
	}

	if(empty(get_option(FYBY_OPTION__METADATAFIELD))) {
		update_option(FYBY_OPTION__METADATAFIELD, 'visible');
	}

	if(empty(get_option(FYBY_OPTION__HISTORYFIELD))) {
		update_option(FYBY_OPTION__HISTORYFIELD, 'visible');
	}

	if(empty(get_option(FYBY_OPTION__ALERTAFTERREQUEST))) {
		update_option(FYBY_OPTION__ALERTAFTERREQUEST, 1);
	}

	if(empty(get_option(FYBY_OPTION__SIDE))) {
		update_option(FYBY_OPTION__SIDE, 'right');
	}

	if(empty(get_option(FYBY_OPTION__ORDER))) {
		update_option(FYBY_OPTION__order, 'description,message,email,explanation,screenshot,metadata,history,termsAccepted,personalDataAccepted,note');
	}
	 
	echo '<div id="feedybacky-container"></div>';
	echo '<script type="text/javascript" src="'.WP_PLUGIN_URL.'/feedybacky/feedybacky/js/feedybacky.min.js"></script>';
		
	$params = [];
	if(!empty(get_option(FYBY_OPTION__LANGUAGE))) {
		$params[] = 'language: "'.get_option(FYBY_OPTION__LANGUAGE).'"';
	}

	if(!empty(get_option(FYBY_OPTION__ONSUBMIT))) {
		$params[] = 'onSubmit: () => { '.get_option(FYBY_OPTION__ONSUBMIT).' }';
	}

	if(!empty(get_option(FYBY_OPTION__ONSUBMITURL))) {
		$params[] = 'onSubmitUrl: "'.get_option(FYBY_OPTION__ONSUBMITURL).'"';
	}
	
	if(!empty(get_option(FYBY_OPTION__APIKEY))) {
		$params[] = 'apiKey: "'.get_option(FYBY_OPTION__APIKEY).'"';
	}
	
	if(!empty(get_option(FYBY_OPTION__PROJECTSYMBOL))) {
		$params[] = 'projectSymbol: "'.get_option(FYBY_OPTION__PROJECTSYMBOL).'"';
	}
	
	if(!empty(get_option(FYBY_OPTION__TERMSURL))) {
		$params[] = 'termsUrl: "'.get_option(FYBY_OPTION__TERMSURL).'"';
	}
	
	if(!empty(get_option(FYBY_OPTION__PRIVACYPOLICYURL))) {
		$params[] = 'privacyPolicyUrl: "'.get_option(FYBY_OPTION__PRIVACYPOLICYURL).'"';
	}

	if(!empty(get_option(FYBY_OPTION__EXTRAINFO))) {
		$params[] = 'extraInfo: () => { return { '.get_option(FYBY_OPTION__EXTRAINFO).' }; }';
	}

	if(!empty(get_option(FYBY_OPTION__EMAILFIELD))) {
		$params[] = 'emailField: true';
	}

	if(!empty(get_option(FYBY_OPTION__SCREENSHOTFIELD))) {
		$params[] = 'screenshotField: "'.get_option(FYBY_OPTION__SCREENSHOTFIELD).'"';
	}

	if(!empty(get_option(FYBY_OPTION__METADATAFIELD))) {
		$params[] = 'metadataField: "'.get_option(FYBY_OPTION__METADATAFIELD).'"';
	}

	if(!empty(get_option(FYBY_OPTION__HISTORYFIELD))) {
		$params[] = 'historyField: "'.get_option(FYBY_OPTION__HISTORYFIELD).'"';
	}

	if(!empty(get_option(FYBY_OPTION__HISTORYLIMIT))) {
		$params[] = 'historyLimit: "'.get_option(FYBY_OPTION__HISTORYLIMIT).'"';
	}
	
	if(!empty(get_option(FYBY_OPTION__URLTRACKING))) {
		$params[] = 'urlTracking: true';
	}
	
	if(!empty(get_option(FYBY_OPTION__URLTRACKINGLIMIT))) {
		$params[] = 'urlTrackingLimit: "'.get_option(FYBY_OPTION__URLTRACKINGLIMIT).'"';
	}

	if(!empty(get_option(FYBY_OPTION__BEFORESUBMIT))) {
		$params[] = 'beforeSubmit: (payload) => { '.get_option(FYBY_OPTION__BEFORESUBMIT).' }';
	}

	if(!empty(get_option(FYBY_OPTION__ONSUBMITURLSUCCESS))) {
		$params[] = 'onSubmitUrlSuccess: (statusCode, response) => { '.get_option(FYBY_OPTION__ONSUBMITURLSUCCESS).' }';
	}

	if(!empty(get_option(FYBY_OPTION__ONSUBMITURLERROR))) {
		$params[] = 'onSubmitUrlError: (statusCode, response) => { '.get_option(FYBY_OPTION__ONSUBMITURLERROR).' }';
	}

	if(!empty(get_option(FYBY_OPTION__PREFIX))) {
		$params[] = 'prefix: "'.get_option(FYBY_OPTION__PREFIX).'"';
	}

	if(!empty(get_option(FYBY_OPTION__ALERTAFTERREQUEST))) {
		$params[] = 'alertAfterRequest: true';
	}

	if(!empty(get_option(FYBY_OPTION__ADBLOCKDETECTED))) {
		$params[] = 'adBlockDetected: "'.get_option(FYBY_OPTION__ADBLOCKDETECTED).'"';
	}

	if(!empty(get_option(FYBY_OPTION__SIDE))) {
		$params[] = 'side: "'.get_option(FYBY_OPTION__SIDE).'"';
	}

	if(!empty(get_option(FYBY_OPTION__ORDER))) {
		$params[] = 'order: "'.get_option(FYBY_OPTION__ORDER).'"';
	}

	if(!empty(get_option(FYBY_OPTION__EXPANDMESSAGELINK))) {
		$params[] = 'expandMessageLink: true';
	}

	if(!empty(get_option(FYBY_OPTION__TEXTS))) {
		$params[] = 'texts: { '.get_option(FYBY_OPTION__TEXTS).' }';
	}

	if(!empty(get_option(FYBY_OPTION__CLASSES))) {
		$params[] = 'classes: { '.get_option(FYBY_OPTION__CLASSES).' }';
	}

	if(!empty(get_option(FYBY_OPTION__THEME))) {
		$params[] = 'theme: "'.get_option(FYBY_OPTION__THEME, 'default').'"';
	}

	if(!empty(get_option(FYBY_OPTION__ALLOWEDTHEMES))) {
		$params[] = 'allowedThemes: [ '.get_option(FYBY_OPTION__ALLOWEDTHEMES, '"default", "dark"').' ]';
	}

	if(!empty(get_option(FYBY_OPTION__CATEGORIES))) {
		$params[] = 'categories: [ '.get_option(FYBY_OPTION__CATEGORIES, '').' ]';
	}

	if(!empty(get_option(FYBY_OPTION__PRIORITYFIELD))) {
		$params[] = 'priorityField: true';
	}

	if(!empty(get_option(FYBY_OPTION__SCREENSHOTMETHOD))) {
		$params[] = 'screenshotMethod: "'.get_option(FYBY_OPTION__SCREENSHOTMETHOD, 'html2canvas').'"';
	}
	
	$paramsString = implode(",\n", $params);
	
	echo '<script>
			var feedybacky = new Feedybacky("feedybacky-container", {
				'.$paramsString.'
			});
		</script>';
}

function fyby__add_assets() {
    wp_register_style('fyby__styles', WP_PLUGIN_URL.'/feedybacky/assets/styles.css');
    wp_enqueue_style('fyby__styles');
}

add_action('admin_menu', 'fyby__setup_menu');
add_action('wp_footer', 'fyby__init_plugin');
add_action('admin_menu', 'fyby__add_assets');
